python/pyspark/mllib/clustering.py

```diff
@@ -38,13 +38,116 @@
 from pyspark.mllib.util import Saveable, Loader, inherit_doc, JavaLoader, JavaSaveable
 from pyspark.streaming import DStream
 
-__all__ = ['KMeansModel', 'KMeans', 'GaussianMixtureModel', 'GaussianMixture',
-           'PowerIterationClusteringModel', 'PowerIterationClustering',
-           'StreamingKMeans', 'StreamingKMeansModel',
+__all__ = ['BisectingKMeansModel', 'BisectingKMeans', 'KMeansModel', 'KMeans',
+           'GaussianMixtureModel', 'GaussianMixture', 'PowerIterationClusteringModel',
+           'PowerIterationClustering', 'StreamingKMeans', 'StreamingKMeansModel',
            'LDA', 'LDAModel']
 
 
 @inherit_doc
+class BisectingKMeansModel(JavaModelWrapper):
+    """
+    .. note:: Experimental
+
+    A clustering model derived from the bisecting k-means method.
+
+    >>> data = array([0.0,0.0, 1.0,1.0, 9.0,8.0, 8.0,9.0]).reshape(4, 2)
+    >>> bskm = BisectingKMeans()
+    >>> model = bskm.train(sc.parallelize(data), k=4)
+    >>> model.predict(array([0.0, 0.0])) == model.predict(array([0.0, 0.0]))
+    True
+    >>> model.k
+    4
+    >>> model.computeCost(array([0.0, 0.0]))
+    0.0
+    >>> model.k == len(model.clusterCenters)
+    True
+    >>> model = bskm.train(sc.parallelize(data), k=2)
+    >>> model.predict(array([0.0, 0.0])) == model.predict(array([1.0, 1.0]))
+    True
+    >>> model.k
+    2
+
+    .. versionadded:: 2.0.0
+    """
+
+    @property
+    @since('2.0.0')
+    def clusterCenters(self):
+        """Get the cluster centers, represented as a list of NumPy arrays."""
+        return [c.toArray() for c in self.call("clusterCenters")]
+
+    @property
+    @since('2.0.0')
+    def k(self):
+        """Get the number of clusters"""
+        return self.call("k")
+
+    @since('2.0.0')
+    def predict(self, x):
+        """
+        Find the cluster to which x belongs in this model.
+
+        :param x: Either the point to determine the cluster for or an RDD of points to determine
+        the clusters for.
+        """
+        if isinstance(x, RDD):
+            return x.map(self.predict(x))
```

Comment:
Ah seems that the JavaModelWraper call method being used won't work on the workers. I'll have to port the predict method over.